quill-core/src/main/scala/io/getquill/MirrorIdiom.scala

```diff
@@ -125,16 +125,20 @@ class MirrorIdiom extends Idiom {
   }
 
   implicit def optionOperationTokenizer(implicit liftTokenizer: Tokenizer[Lift]): Tokenizer[OptionOperation] = Tokenizer[OptionOperation] {
-    case OptionFlatten(ast)              => stmt"${ast.token}.flatten"
-    case OptionGetOrElse(ast, body)      => stmt"${ast.token}.getOrElse(${body.token})"
-    case OptionFlatMap(ast, alias, body) => stmt"${ast.token}.flatMap((${alias.token}) => ${body.token})"
-    case OptionMap(ast, alias, body)     => stmt"${ast.token}.map((${alias.token}) => ${body.token})"
-    case OptionForall(ast, alias, body)  => stmt"${ast.token}.forall((${alias.token}) => ${body.token})"
-    case OptionExists(ast, alias, body)  => stmt"${ast.token}.exists((${alias.token}) => ${body.token})"
-    case OptionContains(ast, body)       => stmt"${ast.token}.contains(${body.token})"
-    case OptionIsEmpty(ast)              => stmt"${ast.token}.isEmpty"
-    case OptionNonEmpty(ast)             => stmt"${ast.token}.nonEmpty"
-    case OptionIsDefined(ast)            => stmt"${ast.token}.isDefined"
+    case UncheckedOptionFlatMap(ast, alias, body) => stmt"${ast.token}.flatMapUnchecked((${alias.token}) => ${body.token})"
```

Comment:
I don't think we should have the `*Unchecked` methods. A better way of doing this would be allowing users to use `option.get`, but I think we should avoid it.

Reply:
I'm concerned with potential performance issues that might be introduced from forcing our users to have case statements in their `where` clauses whereever there are nullable fields. Say that they just want to do:
```scala
case class Person(name:String, age:Option[Int])
query[Person].filter(p => p.age.map(_ + 1).exists(_ > 18))
```
The way it is now it's just:
```sql
select ... from Person p where p.age + 1 > 18
```
Now with this change it would be:
```sql
select ... from Person p where 
((case when p.age is not null then p.age + 1 else null) is not null) &&
((case when p.age is not null then p.age + 1) > 18)
```
This is significantly more complex for a query analyzer to optimize and the query itself is much less readable (Btw, slick currently does something like this!).
Thus, in this particular query, using the old way (i.e. the way it is now) does not pose any risk of non-ANSI behavior in any database and as a user, in certain places I would definitely prefer it.
This is why I introduced the `Unchecked*` methods. 
So that our users can control what happens to their queries.

Reply:
we'd need only `option.get` to achieve that, no?

Reply:
That would work but wouldn't users expect a `NoSuchElementException` to be thrown in many of those cases? Say you have something like this:
```scala
case class Person(name:Option[String])
query[Person].map(n => Some(name.get + " M.D."))
```
Normally, if `p.name` is None, we would expect a `NoSuchElementException` to be thrown but in our case, the value would be None. I think this is pretty inconsistent and might confuse users.

Reply:
`get` is as unsafe as the `*Unsafe` methods, I'd prefer to have it instead of adding new methods to the DSL. Maybe we could call it `unsafeGet`?

Reply:
What about `orNull`? That would be completely consistent with `Option` behavior.

Reply:
Hum, well if we do it in the `valueParser` like we do `Some.apply`, `Option.apply` etc... it's practically trivial. Just do:
```scala
case q"$o.orNull" if is[Option[Any]](o) => astParser(o)
```
We don't need additional AST elements to make it work.

Reply:
The AST should match the original code as much as possible. It's the case that `orNull` has this behavior for SQL but it might not be the case in other target languages. That's why we flatten the options in the sql module

Reply:
Okay. If really you want to go this direction, let's do it right. I suggest we replace this: 
```scala
    case q"scala.None"                   => NullValue
    case q"scala.Option.empty[$t]"       => NullValue
    case q"scala.Some.apply[$t]($v)"     => astParser(v)
    case q"scala.Option.apply[$t]($v)"   => astParser(v)
```
With this:
```scala
    case q"scala.None"                   => OptionNone
    case q"scala.Option.empty[$t]"       => OptionNone // or maybe even have OptionEmpty?
    case q"scala.Some.apply[$t]($v)"     => OptionSome(astParser(v))
    case q"scala.Option.apply[$t]($v)"   => OptionApply(astParser(v))
    case q"$o.orNull[$t]($v)" if is[Option[Any]](o) => OptionOrNull(astParser(o))
```
That way in the future, when we want a typed-AST, we have a decent chance of actually making optionals work properly. Wdyt?

Reply:
Shoot! `orNull` does not work for `Int`, `Double`, etc... How about `orNullValue` or `orEmpty`?